core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala

```diff
@@ -49,8 +49,8 @@ private[spark] class AppClient(
   private val REGISTRATION_TIMEOUT_SECONDS = 20
   private val REGISTRATION_RETRIES = 3
 
-  private var endpoint: RpcEndpointRef = null
-  private var appId: String = null
+  @volatile private var endpoint: RpcEndpointRef = null
```

Comment:
it's better to use an atomic reference to be more explicit here. i'm going to merge this and update it.